### PR TITLE
Fallback to a secondary protocol if primary (https) is not available

### PR DESCRIPTION
**Pull request by ahungry**
This should fix the issue with the browser always defaulting to https by dropping to http when the https version of a site (such as http://ahungry.com or http://ddg.gg) is not available.

However, I think there is an unrelated issue with webkit + Arch Linux, in that if the browser actually does drop to the fallback (by catching the #'handler-case error of usocket) it deadlocks.

Can someone on Ubuntu (or non Arch) test this commit and see if it works?

Expected behaviour is to be able to open just 'ahungry.com' and actually get the site vs a non-existing https://ahungry.com connection.

### CODE REVIEW

**Comment by ahungry** [comment]
The commit I just pushed fixes the issue of failed functionality - I can successfully use the 'https' default scheme/protocol and fall back to 'http' when it is not available with this merge request.

There is probably a miniscule chance of a race condition the way I've implemented it, but I think this commit set makes the browser much more usable.

**Comment by Ralt** [comment]
Chrome's behavior is to fail when going to https://ahungry.com. Shouldn't lispkit behave the same?

Lispkit shouldn't crash though.

**Comment by AeroNotix** [comment]
@Ralt the difference is that typing `ahungry.com` goes to `http` by default, which works, and allows the webpage to redirect to `https`. On the other hand, typing to `https` fails since it won't drop the security of the connection 'silently'.

I'm thinking perhaps a better approach might be to make `http` the default protocol since this is what Chrome does, allowing people to upgrade to `https` with knowing the caveats involved.

Thoughts?

File: browser.lisp [hunk 1]
```diff
@@ -152,7 +152,39 @@
   "Move backwards a page."
   (webkit2:webkit-web-view-go-back (webview browser)))
 
+(defparameter *fallback-timeout* .2)
+
+;; 20150212 - MDC - Fixes issue with unavailable default scheme when opening url
```
Comment (AeroNotix) [inline browser.lisp]: Comments like this are unnecessary, we have VCS for attributing reasoning to code.

**Comment by Ralt** [comment]
Using http by default seems to be the sensible option. If you make this a parameter, it's easily changeable in the lispkitrc.

Still, lispkit shouldn't crash :-)

**Comment by AeroNotix** [comment]
I've not seen anything which indicates it's Lispkit that crashes, so far.

It could be an interaction with the underlying webkit library that's causing it.

File: browser.lisp (hunk 1 above)
Comment (ahungry) [inline browser.lisp]: Granted there is probably a miniscule possibility of it occurring, but what if the project becomes decoupled with git in the future for a superior VCS system?  I'm sure there were past users of svn/bazaar that migrated to git and regretted not having explanatory comments in the codebase.

Granted, the date stamp/author initials are probably not necessary, but some code comments go a long way (as well as for someone new hopping into the project - attempting to 'git blame' every file to see why each line is there is very cumbersome compared to comments).

Thoughts?

**Comment by ahungry** [comment]
I'll try to run some straces today when using the browser to see if I can find where/why it's locking for me.

Definitely easily reproducible under Arch, but that's the price to pay for rolling release (assuming these issues don't manifest as often on the *buntus).

**Comment by AeroNotix** [comment]
Fwiw I run arch so a proper set of repro steps would be appreciated

**Comment by ahungry** [comment]
I have a reproducible crash procedure here (involving the javascript run calls) which includes the strace:

https://github.com/AeroNotix/lispkit/issues/82

The hard lock on URL not found is not as consistent as the lock linked above, but you can reproduce it by checking out the first commit ffa3e99 I added to the pull request (before I did the usocket calls in it's own thread) and then doing the following after starting the browser:

C-x C-f reddit.com (will load https://reddit.com and makes it past my check)
C-x C-f ahungry.com (fails my check and browser then hangs)

Yet, just calling my #'port-available? function directly works without an issue in the REPL (even in the fail case of port 443 being closed on the remote end).

Checking out the master branch also locks me up pretty frequently as soon as I attempt to load an invalid url (try https://ahungry.com enough times to see what I mean).

It's not a segfault/crash, but the browser just sits there and fails to receive any further input whatsoever (and if I tab away/come back, it's a gray background vs showing any rendered DOM).

**Comment by ahungry** [comment]
Ok, please see this program I just put together:

https://github.com/ahungry/puny-browser/blob/master/main.c

I get absolutely no failure or lock up when executing that - it gives the pop up, I click the button on it, and it closes (and the browser remains where it was without graying out or anything).

I've found the issue here though is with cl-webkit2 not lispkit (necessarily), as I can reproduce the failure by doing:
```lisp
(ql:quickload :cl-webkit2)
(load "~/src/lisp/cl-webkit/demo/simple-browser.lisp")
(simple-browser-main)
```

After updating the function in simple-browser.lisp to read as follows:
```lisp
(defun simple-browser-main ()
  "A single-window browser with no keyboard or mouse input.                                                                    
Loads and renders a single web page."
  (gtk:within-main-loop
    (let ((win (make-instance 'gtk:gtk-window))
          (view (make-instance 'webkit2:webkit-web-view)))
      (gobject:g-signal-connect win "destroy"
                                #'(lambda (widget)
                                    (declare (ignore widget))
                                    (gtk:leave-gtk-main)))
      (gtk:gtk-container-add win view)
      (webkit2:webkit-web-view-load-uri view "http://www.example.com")
      (webkit2:webkit-web-view-run-javascript
       view
       "alert(1);"
       (cffi:null-pointer)
       (cffi:null-pointer)
       (cffi:null-pointer))
      (gtk:gtk-widget-show-all win))))

```

**Comment by AeroNotix** [comment]
In total, I'm really not a fan of doing port checks before sending out requests. I'm thinking we should default to using `http` instead of `https` and allowing the browser to follow `https` redirects. Giving the user the ability to override, knowing the caveats.

**Comment by AeroNotix** [comment]
@ahungry does it seem that it's just related to the `alert` box, though? I've tried this before and it seemed that the problem was related to the alert box. Perhaps we need to catch GTK signals from the alert box and handle them cleanly or something?

I'll play with your example today when I get to some better wifi.

**Comment by AeroNotix** [comment]
@ahungry Here is an old issue, I think this is the same thing. https://github.com/AeroNotix/lispkit/issues/8

File: browser.lisp (hunk 1 above)
Comment (AeroNotix) [inline browser.lisp]: If we moved to a different VCS, history would be imported too.

**Comment by ahungry** [comment]
I don't think this needs to remain open.

**Pull request closed without merging**